More "link" to "body" refactoring.

liangfok opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2597)
<!-- Reviewable:end -->

liangfok commented:
+@sherm1 for both levels of review since this is a minor non-functional change.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKiuJTDbZfaG9CqAowt:276302029)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@liangfok, build failed because this changes the API and forces downstream changes that haven't been made. Also, let's make sure @RussTedrake is happy with the change from "link" to "body" before doing this everywhere. Russ?

liangfok commented:
I agree.

FTR, it looks like unit test `test_python_pydrake.test.testPR2IK` failed.

RussTedrake commented:
what was the motivation for the change?  it will break the current "API".

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKk8GB3SSO4MgpvZQu9:-235022864)*
<!-- Sent from Reviewable.io -->

liangfok commented:
There are two primary motivations:

1. Terminology simplicity and consistency. Note the software abstractions `RigidBody` and `RigidBodyTree` both contain the term "body". Intermixing "body" with "link" will only result in confusion and inconsistency.
2. [This comment](https://github.com/RobotLocomotion/drake/pull/2595#issuecomment-227202986) by @sherm1 indicating that the term "link" is something that's only used in the robotics community.

A secondary motivation is we are currently in a pre-release period where the API is less set-in-stone. It would be best to clean up the API as much as possible now before our first 1.0 release.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKkPQu49D_QdMDx4MTb:1262060624)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
i guess i'm ok with it, but i think you have the burden for now of adding back findLink() and finkLinkId() methods which throw a static assert instructing people to update their code to FindBody.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKkYoFfymZyCV21ePDM:975410270)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Sure no problem! Thanks.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKkaHFKv5M9I5TAPDlk:-921491520)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I attempted to add back `RigidBodyTree::findLinkID()` and `RigidBodyTree::findLink()` with `static_assert` lines in their method bodies but couldn't because that results in a compile-time error even if the methods are not called.

I then considered replacing the `static_assert` lines with thrown exceptions, but this is strictly worse than simply not providing the old methods back since it prolongs the time it takes for downstream users to realize that they need to update their code.

I thus removed these deprecated methods again.

If a downstream user compiles their code and it fails because of a missing `RigidBodyTree::findLinkID()` or `RigidBodyTree::findLink()` method, they can consult the CHANGELOG to determine the new methods that replace them.

Perhaps we can institute a more formal deprecation glide path after release 1.0?

---

Review status: 0 of 11 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKooHqhcJ_mDruSLs9O:-1925879894)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Forgot to mention: Another option I considered was to have the old API call the new API after printing a warning to `std::cerr`. However, this is also not as good as just removing the old API methods since it delays when downstream users discover the need to update to the latest API.

On another note, this PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2187/ and is thus ready for another round of review / merge.

---

Review status: 0 of 11 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKpB0PM2On_8be6PYQJ:-292294313)*
<!-- Sent from Reviewable.io -->

wxmerkt commented:
Would it be possible within the remits of the style guide or similar to add a [deprecation comment](https://google.github.io/styleguide/cppguide.html#Deprecation_Comments), use a pragma/trigger a deprecated compiler warning/output an error on runtime and internally call the new API (just saw you added a new comment to that regard). I feel like having a warning/error message shows users they need to act on it (and the deprecation comment when it's going to be removed, e.g. ``to be removed in 1.1``) but doesn't throw them under the bus immediately.

sherm1 commented:
Since the compilers all support deprecation as a compile-time warning (standardized in C++14), we could define a Drake macro to mark the old methods, which would call the new ones. The warnings are only issued when the old methods are used. Here is an example of a functional macro for that purpose:
```
/* C++14 introduces a standard way to mark deprecated declarations. Before
that we can use non-standard compiler hacks. */
#ifndef SWIG
    #if __cplusplus >= 201402L
        /* C++14 */
        #define DEPRECATED_14(MSG) [[deprecated(MSG)]]
    #elif _MSC_VER
        /* VC++ just says warning C4996 so add "DEPRECATED" to the message. */
        #define DEPRECATED_14(MSG) __declspec(deprecated("DEPRECATED: " MSG))
    #else /* gcc or clang */
        #define DEPRECATED_14(MSG) __attribute__((deprecated(MSG)))
    #endif
#else /* Swigging */
    #define DEPRECATED_14(MSG)
#endif
```

---

Review status: 0 of 11 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKpNJjHtDI7VKAj-_SI:231188057)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
FWIW -- I am open to the idea of using `deprecated` markers and enabling a soft cut-over, but I am unclear in which cases a soft cut-over is required / desired.  On `master` recently, we have already renamed dozens of APIs, shuffled responsibility between classes, hidden public fields behind getters, etc.  Is there a heuristic that guides when we should be doing a soft-transition and when we can do hard?  Prior @RussTedrake's upthread comment in this PR, my understanding was that only matlab APIs are sacred, and C++ APIs can be changed with abandon.

liangfok commented:
Thanks for all of the advice!

How about let's strike a compromise: From now till the 1.0 release, we'll do everything possible to honor the existing method API alongside the new API but add deprecation compile-time and run-time warnings stating that the old APIs will be removed by the 1.0 release. Just prior to the 1.0 release, we remove all of these depreciated methods.

**Caveat 1:** Drake also uses numerous public variables that we should ideally make private or protected. I see no way of doing this without breaking the API, and don' t think we can keep them public until the last minute. For these, I can think of no other option but to force downstream users to use the soon-to-be-introduced accessors for them.

** Caveat 2:** We are considering changing entire class names prior to the 1.0 release. If a class' name is changed, I see no way to maintain both simultaneously unless the new classes are decoupled from the old ones and ideally physically located in a separate directory. This may justify developing the new classes in a separate directory alongside the existing classes (e.g., System 2.0 vs. System 1.0 and potentially Dynamics 2.0 vs. Dynamics 1.0).

Thoughts?

RussTedrake commented:
I agree that we should be able to change the c++ code with abandon.  I'll go further -- i think it's really important that we feel unconstrained in making those changes.  However, findLink is a method that probably every piece of c++ code using drake is calling. Changing it somewhat arbitrarily (because we suddenly like FindBody better) feels like an extreme of abuse to any users, and i hoped that adding a deprecated warning is very easy in this case.  If there is an easy solution available, then i think we should consider it.

liangfok commented:
Help! I need a C++ guru to tell me what's what's wrong with the way I'm marking a method as deprecated:

[Here is the code in question.](https://github.com/liangfok/drake/blob/4e976fffb109e4f9c533d0981482b2466d21b4aa/drake/systems/plants/RigidBodyTree.h#L582)

Here's the error I'm getting using `clang` on OS X:
```
/Users/liang/dev/drake/drake/../drake/systems/plants/RigidBodyTree.h:582: Error: Syntax error - possibly a missing semicolon.
```
I basically copy / pasted an example from here:
http://clang.llvm.org/docs/LanguageExtensions.html#messages-on-deprecated-and-unavailable-attributes

Version of compiler:
```
$ clang --version
Apple LLVM version 7.3.0 (clang-703.0.29)
Target: x86_64-apple-darwin15.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Here's a small program that @rpoyner-tri gave me, which I used to verify that `deprecate` is supported by my compiler:

```
#if __has_extension(attribute_deprecated_with_message)
#error "deprecated with message OK"
#else
# error "deprecated with message UNSUPPORTED"
#endif
```

The output indicates it's supported by my compiler:

```
$ clang -c test_deprecated.cc
test_deprecated.cc:2:2: error: "deprecated with message OK"
#error "deprecated with message OK"
 ^
1 error generated.
```

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKt_X2Dtn8GxXsC6_S2:554122538)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
`#error` always generates a compilation error when it is encountered by the compiler.
I'm going to submit a `DRAKE_DEPRECATED` macro PR shortly.

rpoyner-tri commented:
Just to be clear, the #error machinery was just a temporary tactic for diagnosing the compiler problem. We resolved it f2f. I look forward to your DEPRECATED PR.

liangfok commented:
Latest update: @rpoyner-tri helped me resolve the C++ issues but I'm unfortunately still running into a `SWIG` issue. Here is the error:

```
[205/216] Swig source
FAILED: bindings/swig/rbtreePYTHON_wrap.cxx bindings/swig/rbtree.py
cd /Users/liang/dev/drake-build/drake/bindings/swig && /usr/local/Cellar/cmake/3.5.0/bin/cmake -E make_directory /Users/liang/dev/drake-build/drake/bindings/swig && /Users/liang/dev/drake-build/install/bin/swig -python -w315,-401 -w503 -I/Users/liang/dev/drake-build/install/share/swig/python -I/Users/liang/dev/drake-build/install/share/swig/common -modern -outdir /Users/liang/dev/drake-build/drake/bindings/swig -c++ -I/Users/liang/dev/drake-build/drake/generated -I/Users/liang/dev/drake/drake/.. -I/Users/liang/dev/drake-build/drake/exports -I/Users/liang/dev/drake-build/install/include -I/Users/liang/dev/drake-build/drake/lcmgen -I/usr/local/Cellar/glib/2.46.2/include/glib-2.0 -I/usr/local/Cellar/glib/2.46.2/lib/glib-2.0/include -I/usr/local/opt/gettext/include -I/Users/liang/dev/drake-build/install/include/eigen3 -I/Users/liang/dev/drake/drake/thirdParty/spruce/include -I/Users/liang/dev/drake/drake/thirdParty/cimg -I/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/include/python2.7 -I/usr/local/lib/python2.7/site-packages/numpy/core/include -o /Users/liang/dev/drake-build/drake/bindings/swig/rbtreePYTHON_wrap.cxx /Users/liang/dev/drake/drake/bindings/swig/rbtree.i
/Users/liang/dev/drake/drake/../drake/systems/plants/RigidBody.h:215: Warning 325: Nested class not currently supported (CollisionElement ignored)
/Users/liang/dev/drake/drake/../drake/systems/plants/RigidBodyTree.h:583: Error: Syntax error - possibly a missing semicolon.
ninja: build stopped: subcommand failed.
```

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKttF37FjNc7_63sdxB:1106906802)*
<!-- Sent from Reviewable.io -->

liangfok commented:
The latest push includes a workaround that compiles on OSX. Will look into how to remove the work unless somebody else beats me to it.

liangfok commented:
I've tried everything I can think of to remove the workaround but haven't found a solution. @sherm1 I hope your pending PR containing the deprecation macro is able to solve this problem. FYI, the workaround shows up here:

https://github.com/liangfok/drake/blob/feature/link_to_body_refactor/drake/systems/plants/RigidBodyTree.h#L584

Without those `#ifndef SWIG` guards that are external to the header in the `.h` file, I'm getting the following compile-time error:

```
/Users/liang/dev/drake/drake/../drake/systems/plants/RigidBodyTree.h:585: Error: Syntax error - possibly a missing semicolon.
```

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKuReB20pdv8IYwe1nj:308133897)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
The `DEPRECATED` macro comes *before* the deprecated class or function.

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKuZ-Xq6L4XdyxOegsp:-621438876)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I just tried the following combinations to no avail:

```
   DEPRECATED_14("BOO")
   RigidBody* findLink(const std::string& link_name,
                       const std::string& model_name = "",
                       int model_id = -1) const;
```
I'm wondering if there's a way to tell what SWIG gets after the preprocessing stage.

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKuc7wFQ4JyADl-QNzx:-1846585272)*
<!-- Sent from Reviewable.io -->

liangfok commented:
The original workaround did not work on Windows. I refactored it and also merged in #2621. Hoping this will now pass CI.

---

Review status: 0 of 16 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKxg4Xp8BH2OrBxrc5_:2129632570)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2240/ and is thus ready for another round of review / merge.

---

Review status: 0 of 16 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKy-jKALqIuUU69T0tW:86220141)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2251/ and is thus ready for final review / merge. +@jwnimmer-tri 

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKyppx5shY6_SA1jYIx:-2031014011)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Shouldn't this be merged or rebased to master, to use the existing deprecated.h instead of a different one?  Both reviewable and github are showing drake/common/deprecated.h as new-code in this PR.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKyr-ROH1iDlsCNE96Y:2005706377)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I think that may be a Reviewable issue. I'll close and re-open this PR to fix.

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKyrD4evIp8B1wBSTrr:-586684062)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Looks like this includes your own `deprecated.h`. The master branch has `drake_deprecated.h` now so you should delete yours.

---

Reviewed 1 of 12 files at r9.
Review status: 1 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[CHANGELOG.md, line 79 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyrzq_a9GaWknMiT2h:-KKyrzq_a9GaWknMiT2i:1319725140) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/CHANGELOG.md#L79)):*
> ```Markdown
> [#2597]: https://github.com/RobotLocomotion/drake/issues/2597
> [#2602]: https://github.com/RobotLocomotion/drake/issues/2602
> [#2621]: https://github.com/RobotLocomotion/drake/issues/2621
> ```

The PR # is wrong and there is no final newline.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKysCV5gA6rjr8M8xNE:-1681105336)*
<!-- Sent from Reviewable.io -->

liangfok commented:
You're right! Doing that right now. Thanks.

---

Reviewed 1 of 12 files at r9.
Review status: 2 of 12 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KKysdy0GjEDi6zvcxvQ:1796165750)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 2 of 12 files at r9.
Review status: 2 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 10 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyszJVISI_kzY2gyU1:-KKyszJVISI_kzY2gyU2:1200672934) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L10)):*
> ```objc
> #include <unordered_map>
> 
> #include "drake/common/deprecated.h"
> ```

`drake_deprecated.h`

---

*[drake/systems/plants/RigidBodyTree.h, line 581 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyshnUyy05WVVvAuQd:-KKyshnUyy05WVVvAuQe:-55963613) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L581)):*
> ```objc
>    * instead.
>    */
> #ifndef SWIG
> ```

@liangfok, this shouldn't be necessary since `DRAKE_DEPRECATED` has its own `#ifdef SWIG`. What happens if you remove these now?

---

*[drake/systems/plants/RigidBodyTree.h, line 584 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKytDRI4MwiAg6Sqx_N:-KKytDRI4MwiAg6Sqx_O:-1733750330) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L584)):*
> ```objc
>       DRAKE_DEPRECATED(
>           "RigidBodyTree: findLink: WARNING: This method is "
>           "deprecated. Please use the newer "
> ```

This message is wrong -- the macro generates its own boilerplate and the compiler will identify the deprecated declaration. You should just write `DRAKE_DEPRECATED("Please use RigidBodyTree::FindBody instead.")`.

---

*[drake/systems/plants/RigidBodyTree.h, line 616 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKytTgR4dYAunaUTukt:-KKytTgSDO4UDgZ_DRDN:-2081937833) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L616)):*
> ```objc
>       DRAKE_DEPRECATED(
>           "RigidBodyTree: findLinkId: ERROR: This "
>           "method is deprecated. Please use the newer "
> ```

Fix message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[CHANGELOG.md, line 79 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyrzq_a9GaWknMiT2h:-KKyt47flY0f6YCI3wUX:-795674652) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/CHANGELOG.md#L79)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> The PR # is wrong and there is no final newline.
</details>
Fixed. Thanks.

---

*[drake/systems/plants/RigidBodyTree.h, line 10 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyszJVISI_kzY2gyU1:-KKyuAQzOqL7ewqgvuQU:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L10)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> `drake_deprecated.h`
</details>
Fixed.

---

*[drake/systems/plants/RigidBodyTree.h, line 581 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyshnUyy05WVVvAuQd:-KKyuCszVgQ3XqcZrAoB:-1288540263) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L581)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> @liangfok, this shouldn't be necessary since `DRAKE_DEPRECATED` has its own `#ifdef SWIG`. What happens if you remove these now?
</details>
The following error occurred yesterday:

```
Error: Syntax error - possibly a missing semicolon.
```
I tried a whole bunch of things to avoid this work around but they all failed.

I am re-testing right now using the latest version of this PR's branch.

---

*[drake/systems/plants/RigidBodyTree.h, line 584 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKytDRI4MwiAg6Sqx_N:-KKyuo0MN41vRUZnI9X0:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L584)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This message is wrong -- the macro generates its own boilerplate and the compiler will identify the deprecated declaration. You should just write `DRAKE_DEPRECATED("Please use RigidBodyTree::FindBody instead.")`.
</details>
Done.

---

*[drake/systems/plants/RigidBodyTree.h, line 616 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKytTgR4dYAunaUTukt:-KKyup-lA8wJb8BE9Q4A:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L616)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Fix message.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 5 of 12 files at r9, 1 of 1 files at r11.
Review status: 8 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[CHANGELOG.md, line 79 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyrzq_a9GaWknMiT2h:-KKyyTEJhOI-TK5jCyze:1068786568) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/CHANGELOG.md#L79)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Fixed. Thanks.
</details>
still no final newline

---

*[drake/systems/plants/RigidBodyTree.h, line 581 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyshnUyy05WVVvAuQd:-KKyz7cYlR52MWNoiEUm:641171266) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L581)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> The following error occurred yesterday:
> 
> ```
> Error: Syntax error - possibly a missing semicolon.
> ```
> I tried a whole bunch of things to avoid this work around but they all failed.
> 
> I am re-testing right now using the latest version of this PR's branch.
</details>
I wonder if SWIG has to be told explicitly which macros to expand? If it did expand this one it would have encountered the internal `#ifndef SWIG` that would have made it expand to nothing.

---

*[drake/systems/plants/test/rigid_body_tree_test.cc, line 195 \[r13\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKz0eiVJjte2n6IoY71:-KKz0eiVJjte2n6IoY72:311338566) ([raw file](https://github.com/robotlocomotion/drake/blob/3495c54e3ca9371a655b0b36da2c2bc25c55762b/drake/systems/plants/test/rigid_body_tree_test.cc#L195)):*
> ```cpp
> //   EXPECT_NE(tree->findLink("body1", "robot1"), nullptr);
> //   EXPECT_NE(tree->findLinkId("body1"), -1);
> // }
> ```

why is this here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 7 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[CHANGELOG.md, line 79 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyrzq_a9GaWknMiT2h:-KKz2GSh4A9TZEkTsnyD:-1542061128) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/CHANGELOG.md#L79)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> still no final newline
</details>
Done. Hmm, my text editor must have removed it because I'm pretty sure I had left one previously. Will look into it later.

---

*[drake/systems/plants/RigidBodyTree.h, line 581 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyshnUyy05WVVvAuQd:-KKz1P6iAFZOIzN2575A:604437515) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L581)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I wonder if SWIG has to be told explicitly which macros to expand? If it did expand this one it would have encountered the internal `#ifndef SWIG` that would have made it expand to nothing.
</details>
I removed the `#ifndef SWIG` and `#endif` preprocessor directives surrounding the `DRAKE_DEPRECATED` macro and encountered the following compile-time error:

```
/home/liang/dev/drake2-distro/drake/../drake/systems/plants/RigidBodyTree.h:582: Error: Syntax error in input(3).
```
How does one tell SWIG which macros to expand?

---

*[drake/systems/plants/test/rigid_body_tree_test.cc, line 195 \[r13\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKz0eiVJjte2n6IoY71:-KKz1rpnAC2i-vu5xlpa:1900975928) ([raw file](https://github.com/robotlocomotion/drake/blob/3495c54e3ca9371a655b0b36da2c2bc25c55762b/drake/systems/plants/test/rigid_body_tree_test.cc#L195)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> why is this here?
</details>
Oops. It was test code I used to evaluate the deprecate macro. Removed in next revision.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 2 of 12 files at r9, 1 of 1 files at r13, 2 of 2 files at r14.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 581 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyshnUyy05WVVvAuQd:-KKz5-qWNC__Dt1wcEEr:470113135) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L581)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I removed the `#ifndef SWIG` and `#endif` preprocessor directives surrounding the `DRAKE_DEPRECATED` macro and encountered the following compile-time error:
> 
> ```
> /home/liang/dev/drake2-distro/drake/../drake/systems/plants/RigidBodyTree.h:582: Error: Syntax error in input(3).
> ```
> How does one tell SWIG which macros to expand?
</details>
SWIG's macro documentation [here](http://www.swig.org/Doc1.3/Preprocessor.html) makes me think it should have no trouble with this at all. I suspect that SWIG is not seeing the `drake_deprecated.h` header so just doesn't know this is a macro and instead thinks it is a broken function call.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 581 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyshnUyy05WVVvAuQd:-KKz72cO3OlCbwDx1Q-V:-61270719) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L581)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> SWIG's macro documentation [here](http://www.swig.org/Doc1.3/Preprocessor.html) makes me think it should have no trouble with this at all. I suspect that SWIG is not seeing the `drake_deprecated.h` header so just doesn't know this is a macro and instead thinks it is a broken function call.
</details>
I see. Makes sense. At @jwnimmer-tri's suggestion, I'm planning on writing a minimal test that replicates this problem.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
> This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2251/ and is thus ready for final review / merge. @jwnimmer-tri

I'm going to de-assign myself.  By all means if merging ends up blocked on a flake, please bring ping me here or DM me on slack to use my override powers, but in general I think either the platform reviewer or author should push the merge button; the on-call reviewer doesn't need to be involved (and will often lack the necessary context).

liangfok commented:
[OS X CI](https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=40617) appears to have suffered a [known flake](https://github.com/RobotLocomotion/drake/issues/2385):

```
ld: malformed archive TOC entry for I[NON-UTF-8-BYTE-0x8B][NON-UTF-8-BYTE-0xBC]$[NON-UTF-8-BYTE-0xB8], offset 274736 is beyond end of file 12288
 file '/Users/yosemite/workspace/experimental/5403bcc7/build/lib/liblcmtypes_bot2-core.a' for architecture x86_64
```

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KL14GCYX44vrZu9adOa:1594070483)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot retest this please

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2276/ and is thus ready for final review / merge.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KL4QSg4Xk2EzJ797SJr:617360650)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Just one SWIG question remaining.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 581 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyshnUyy05WVVvAuQd:-KL4xuU2_woZhZq-8hnP:-1193412551) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L581)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I see. Makes sense. At @jwnimmer-tri's suggestion, I'm planning on writing a minimal test that replicates this problem.
</details>
Liang, there is still this SWIG question open. Do you want to leave it here and deal with it elsewhere? If so please file an issue so we can keep track.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KL4y80YZadyehxjCxQX:219702653)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 581 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-KKyshnUyy05WVVvAuQd:-KL6KD2QDMTPboZH4od-:-807035108) ([raw file](https://github.com/robotlocomotion/drake/blob/a82963b9023f7c866333cfd4a4840ffa19320d8c/drake/systems/plants/RigidBodyTree.h#L581)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Liang, there is still this SWIG question open. Do you want to leave it here and deal with it elsewhere? If so please file an issue so we can keep track.
</details>
Done: https://github.com/RobotLocomotion/drake/issues/2646

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KL6KFnfmvFfo9YCK514:701515998)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KL7vslQaSjWTZP32Ech:1417088721)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Liang, reviewable completion is held up on two of your own comments that need acknowledging.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2597#-:-KL7w-IcGSSp_KZTS36A:-1824837203)*
<!-- Sent from Reviewable.io -->
